feat: add dijkstra pathing

phantomsc2 opened the pull request:
This PR adds a C implementation of Dijkstra's algorithm on grids using a [d-ary heap](https://en.wikipedia.org/wiki/D-ary_heap)

The main difference to the A*-pathing in [MapAnalyzer.astar_path](std::priority_queue) is that paths are precalculated once (per frame) and for all cells on the map. Therefore it is much slower than single-source / single-target path finding and no heuristic is used. However if many units reuse the same grid (say 100 Zerglings) and would follow similar paths, this is much faster because the calculcation is shared.

Technically the implementation works backwards, initializating the priority queue with the targets and storing pointers to predecessor cells until the queue is empty. These pointers turn into forward pointers from any source and are simply followed to construct the final path.

### Notes
- multiple targets: any number of targets can be passed (they can be weighted indirectly by setting the cost array at target points appropriately)
- multiple sources: exhaustive search of all cells from which a target can be reached
- infinite cost marks unpathable cells
- if no path is found, a singleton array with the source point is returned
- cost array is padded with infinite cost border such that no bounds checks are needed

### Example Usage

```py
targets = np.array([u.position.rounded for u in bot.enemy_units], np.intp)
cost = np.where(bot.game_info.pathing_grid.data_numpy.T == 1, 1.0, np.inf).astype(np.float64)
pathing = cy_dijkstra(cost, targets)

for unit in bot.units:
    path = pathing.get_path(unit.position.rounded, limit=7)  # path limit is optional
    unit.move(Point2(path[-1]))
```

Commit:
feat: add dijkstra pathing

Commit:
fix: use python ints as points

Commit:
test: add dijkstra testing

Commit:
doc: add docstrings to cy_dijkstra

Commit:
test: slightly adapt test cases

Commit:
fix: optimize bounds checks

Commit:
test: use approximate checks

Commit:
fix: remove unnecessary import

raspersc2 commented:
Thanks for this!

Could you include a `dijkstra.pyi` file? You can look at the existing `pyi` files for reference. This should include the method signature with type information.
This allows intellisense and typechecking to work and prevents some IDEs reporting import errors

Example:
![image](https://github.com/user-attachments/assets/88bd86ca-81d7-4e81-9a72-5a8f77d2b06b)


Furthermore you can add documentation for the autodocs to pick up, you can include this file in `docs/index.md` so it shows up in the docs when this is merged.

I will investigate the build errors

Commit:
fix: Add dijkstra.pyi

Commit:
fix: add dijkstra output members to pyi

Commit:
fix: refactoring

Commit:
test: simplify dijkstra fail case tests

Commit:
fix: use int datatype

Commit:
doc: add dijkstra

phantomsc2 commented:
Added the docs and `dijkstra.pyi`, intellisense is now working inPyCharm.

Also added a small notebook where you can add targets on a map and see the pathing on mouse-over. Up to you if you want to include it.

![image](https://github.com/user-attachments/assets/005fb4f9-f01f-4506-9efa-83212c548d3c)

Commit:
fix: typo in error message

Commit:
fix: set c++11 compiler flag

phantomsc2 commented:
macos wheels seem to be fixed with `extra_compile_args=["-std=c++11"]`

Commit:
feat: add priority queue

Commit:
fix: optimize and unroll

phantomsc2 committed:
feat: add priority queue

* feat: add priority queue

* fix: optimize and unroll

---------

Co-authored-by: Volker Schmidt <volker.schmidt91@gmail.com>

Commit:
ci: revert change to file deletion

Commit:
Merge branch 'dev'

Commit:
ci: revert change to file deletion

Commit:
fix: revert changes to build.py

Commit:
fix: pyproject.toml

Commit:
0.1.0

Automatically generated by python-semantic-release

Commit:
Revert "0.1.0"

This reverts commit 1b60b47a3af721dbcaae5a717270829f0acac6e6.

Commit:
docs: improve dijkstra.pyi
